### ISSUE TITLE
Naming of the `GradValues` function

### ISSUE THREAD

**Issue by jli05** [google/objax#40]
If I understand right, `GradValues` essentially does two things: computing gradients and computing model final values.

So why not split it into two functions? Or if we keep the current form, could we name it `GradAndValuesFn`? Just thinking this is a prominent function and want to keep it the easiest for people beginning to use the framework. An easy name as `fit()` and `predict()` made `scikit-learn`.

**Comment by AlexeyKurakin** [comment]
Yes, `GradValues` returns both gradient and value of the input function/module.

It's pretty much Objax wrapper for [jax.value_and_grad](https://jax.readthedocs.io/en/latest/jax.html#jax.value_and_grad) function. This is where naming comes from.

The reason why this module returns both gradient and value for following reasons:

1) To compute gradient, forward pass has to be performed. Thus value of the function is available for free.
2) Quite often, user need both value and gradient. Example: when compute both value and gradient of loss function during training. In such case returning both gradient and value is a performance optimization, compared to computing gradient and value separately.

As an option we can add module `Grad` which will only returns gradient.
`Grad` could be implemented either as a wrapper for [jax.grad](https://jax.readthedocs.io/en/latest/jax.html#jax.grad) or by calling `GradValues` and throwing away value of the function.
In both cases (wrapper of `jax.grad` or using `GradValues` internally), `Grad` will be computing both value and gradient and throwing away value of the function.

Let us know if module `Grad` will be useful and make things more clear.
Also if you willing to contribute, feel free to implement it and submit a pull request.
Implementation of [GradValues](https://github.com/google/objax/blob/ddcdd30e00575dd2bfcffa988ea7b4388bd1ff8a/objax/gradient.py#L26) could be used as a reference.

**Comment by sathish-a** [comment]
@jli05 thanks for bringing up this. @AlexeyKurakin I've implemented the `Grad` by calling `GradValues` internally and returned only the gradients. I tested it with MNIST code in the examples just to check If I'm affecting any other flows and it worked fine. Now that I couldn't find any guidelines about what else to be added like (creating .rst, etc.) if a new module is introduced. Can you help me with it?

**Comment by jli05** [comment]
Why the forward function valuation is "needed" for computing the gradients whilst mathematically it doesn't necessarily look so?

I wrote the question:

1. Engineering-wise, it seems better to split a function which does two things into two that does one each, so that we give choice, esp. the value part can already be given by `model(input)` thus may seem "redundant".

2. Imagine we teach AI to 2nd grader, the teacher can abstract away the training part, only later on reveal

``` python
def train(model, input, output, ...):
    grad_fn = grad(model)
    for j in range(epochs):
        input_batch = ...
        params = params - learning_rate * grad_fn(input_batch)
```

For me it is cleaner conceptually. If internally we can indeed save computing model value, it is also more efficient.

3. If we choose to do in one go grad and value, I'd think at least we add the word "and" in the function name, just like JAX.

I tend to think these thoughts may occur naturally to anyone who's a new or old learner and a framework should be as simple, little-code as possible to afford best flexibility and many qualities.

**Comment by AlexeyKurakin** [comment]
@jli05 
Most (or maybe all) of the modern deep learning libraries are using back-propagation algorithm to compute gradients. Back propagation need to computes forward pass (i.e. value of the function) to be able to compute gradient.

For example, let's say you have feed-forward neural network which is expressed as a complex function:

```
f(x) = f1( f2( f3( f4(x) ) ) )
```

The gradient of this function is expressed as following:

```
f'(x) = f1'( f2( f3( f4(x) ) ) ) * f2'( f3( f4(x) ) ) * f3'( f4(x) ) * f4'(x)
```

where `f'(x)` is a gradient of function `f` in point `x`.

So in order to compute gradient of `f(x)` value of `f2( f3( f4(x) ) )` has to be computed as well.
In terms of computational complexity, computing value `f2( f3( f4(x) ) )` is different from computing `f(x)` by only one operation `f1`. For large neural network which are used in practice, there is very little difference between computing `f(x)` and `f2( f3( ... ) )`.

You can read more details about backrpopagation [in wikipedia](https://en.wikipedia.org/wiki/Backpropagation) or [in section 6.5 of Deep Learning book](https://www.deeplearningbook.org/contents/mlp.html).

Nevertheless, I agree that it's a good idea to add `Grad` module which only computes gradients, because value of the function might be not always needed.

I'm not so sure whether `GradAndValues` will be better than `GradValues`, but we will discuss it with the team as well.

**Comment by AlexeyKurakin** [comment]
@sathish-a Thanks for volunteering to help implement `Grad` module.

You can add this module into [objax/gradient.py](https://github.com/google/objax/blob/master/objax/gradient.py). Follow existing code of `GradValues` as an example of how to add module.
Just make sure to add unit tests into [tests/gradient.py](https://github.com/google/objax/blob/master/tests/gradient.py).

Most of the documentation is autogenerated from source code, RST files are only needed for customization. In this case it's enough to simply add proper docstring comments to the code of the `Grad` module.

Before creating pull request make sure that all tests are passing and that there are no linter errors. See https://github.com/google/objax#running-linter-and-tests for instructions of how to run unit tests and linter. Also see https://github.com/google/objax#developer-installation-guide for general developer instructions.

**Comment by AlexeyKurakin** [comment]
Grad module was added, so closing this issue.
